File: python/patterns/frequency_counter.py

# sameArraySquared should check if two array have the correspoding squared number
def sameArraySquared(arrOne, arrTwo):
    # check if the length of the arrays are same
    if len(arrOne) != len(arrTwo):
        return False

    # Make a dict for each arrays to save the frequency
    freqArrayOne = {}
    freqArrayTwo = {}

    # Interate thru the first array and save the frequency
    for key in arrOne:
        if key in freqArrayOne:
            freqArrayOne[key] += 1
        else:
            freqArrayOne[key] = 1

    # Interate thru the second array and save the frequency
    for key in arrTwo:
        if key in freqArrayTwo:
            freqArrayTwo[key] += 1
        else:
            freqArrayTwo[key] = 1

    # Interate thru the first frequency dict
    for key in freqArrayOne:
        # Check if the key squared is on second array frequency
        if key ** 2 not in freqArrayTwo:
            return False

            # Check if the value is same
        if freqArrayTwo[key ** 2] != freqArrayOne[key]:
            return False

    return True

File: javascript/src/patterns/frequency-counter.ts
/**
 * sameArraySquared gets two array and compares if the second array represents the square number of the first array
 * Time complexity: O(n)
 * */
function sameArraySquared(firstArray: number[], secondArray: number[]): boolean {
    // check if the array have same length
    if (firstArray.length !== secondArray.length) {
        return false
    }

    // create object for save frequency of each array
    const freqCounterOne: Record<string, number> = {}
    const freqCounterTwo: Record<string, number> = {}

    // iterate on first array to get the frequency
    for (const key of firstArray) {
        freqCounterOne[key] = (freqCounterOne[key] || 0) + 1
    }

    // iterate on second array to get the frequency
    for (const key of secondArray) {
        freqCounterTwo[key] = (freqCounterTwo[key] || 0) + 1
    }

    // iterate on first array and compares the frequency, if mismatch return false
    for (const key in freqCounterOne) {
        if (!(Number(key) ** 2 in freqCounterTwo)) {
            return false
        }
        if (freqCounterTwo[Number(key) ** 2] !== freqCounterOne[key]) {
            return false
        }
    }

    return true
}

export { sameArraySquared }
